### PR TITLE
Addrspace basics

### PR DESCRIPTION
**Pull request by fbs**
To correctly read data on architectures with overlapping address spaces `probe_read` got split up in a `_kernel` and a `_user` variant. As we're responsible for selecting the right helper we have to keep track of the address space a pointer is in.

This PR implements the basics  required to do so, it;

- updates the pointer type to support address space information
- extends the probe_read irbuilder methods to select the right helper based on the addrspace info passed
- extends the ast printer to include addrspace info
- adds 2 new calls, kptr() and uptr() that change the addrspace of a pointer. This is left undocumented for now.

I initially tried to implement this in one go (#1141) but that had some corner cases that where hard to fix. Getting the basics in first makes reviewing easier and makes it easier for others to help out with it.

---

### warning

as this depends on a new kernel  the new behaviour won't be covered by the runtime tests in the ci

### CODE REVIEW

**Comment by fbs** [comment]
ready for review  @sumanthkorikkar, @danobi , @mmisono

**Comment by sumanthkorikkar** [comment]
Will check. Thanks

File: src/ast/irbuilderbpf.h [hunk 1]
```diff
@@ -96,9 +165,16 @@ class IRBuilderBPF : public IRBuilder<>
...
+                                const location &loc);
   CallInst   *createMapLookup(int mapfd, AllocaInst *key);
-  Constant   *createProbeReadStrFn(llvm::Type * dst, llvm::Type * src);
+  Constant *createProbeReadStrFn(llvm::Type *dst,
+                                 llvm::Type *src,
+                                 AddrSpace as);
+  libbpf::bpf_func_id selectProbeReadHelper(AddrSpace as, bool str);
 
   std::map<std::string, StructType *> structs_;
   // clang-format on
```
Comment (danobi) [inline src/ast/irbuilderbpf.h:180]: Delete this too?

File: src/ast/codegen_llvm.cpp [hunk 2]
```diff
@@ -985,6 +985,11 @@ void CodegenLLVM::visit(Call &call)
     expr_ = b_.CreateIntCast(expr_, b_.getInt64Ty(), arg.type.IsSigned());
     b_.CreateOverrideReturn(ctx_, expr_);
   }
+  else if (call.func == "kptr" || call.func == "uptr")
+  {
+    auto &arg = *call.vargs->at(0);
+    arg.accept(*this);
```
Comment (danobi) [inline src/ast/codegen_llvm.cpp:991]:
```suggestion
    auto arg = call.vargs->at(0);
    accept(arg);
```

**Comment by sumanthkorikkar** [comment]
In my opinion, Having runtime-tests would be great for all platforms. May be we should work on the subsequent commits after this integration.

**Comment by mmisono** [comment]
It might be good to add FieldAccess support (e.g., `kptr(args)->x`) to this PR. Otherwise, we cannot do any useful things.

**Comment by fbs** [comment]
> In #614 the plan was to have default addr spaces for pointers in certain probes (ie kprobe ptr args are kernel, uprobe ptr args are user, etc). Is the plan still to do that? We might be able to get away with not doing that and let sparc/ppc/s390x use the kptr and uptr calls.

My concern is that if we don't do it bpftrace would be CPU dependent, people would have to maintain patched tools and it would make non-x86 archs even more a 2nd class citizen.

Overal I don't think it will be *that* hard to properly implement it. The original PR suffered from the type system, it should be easier now. 

> In my opinion, Having runtime-tests would be great for all platforms. May be we should work on the subsequent commits after this integration.

I'll add some basic tests, even if though they don't run in the CI they will be useful.

> It might be good to add FieldAccess support (e.g., `kptr(args)->x`) to this PR. Otherwise, we cannot do any useful things.

Hmm ill look into it

**Comment by danobi** [comment]
> Overal I don't think it will be that hard to properly implement it. The original PR suffered from the type system, it should be easier now.

Just to confirm: this PR doesn't implement it but the plan is to follow up with it, right?

**Comment by danobi** [comment]
>> In my opinion, Having runtime-tests would be great for all platforms. May be we should work on the subsequent commits after this integration.
>>
> I'll add some basic tests, even if though they don't run in the CI they will be useful.

@sumanthkorikkar I think GH actions supports hosted runners ( https://docs.github.com/en/actions/hosting-your-own-runners/about-self-hosted-runners ). This could be one way to get additional architecture support for the CI.

I don't think any of us mind if you come and fix regressions after they occur but you could probably save yourself work in the long term if we can find a way to upstream your CI setup.

**Comment by fbs** [comment]
> > Overal I don't think it will be that hard to properly implement it. The original PR suffered from the type system, it should be easier now.
> 
> Just to confirm: this PR doesn't implement it but the plan is to follow up with it, right?

Yes this just the basics framework to keep track of the AS and to get some agreement on how it should look.

**Comment by sumanthkorikkar** [comment]
Hi @fbs,

Few things:
- Tested few cases on s390x. kptr() and uptr() resolves to bpf_probe_read_kernel() and bpf_probe_read_user.
- -v does not seems to show the bpf bytecode. It is shown only during error. This is not relevant to this PR. However, It makes me to use bpftool each time to see the bpf bytecode.
- Tried to change bashreadline.bt 
    * printf("%-6d %s\n", pid, str(uptr(retval)));
    * ERROR: uptr() only supports pointer arguments (integer provided)
    *  printf("%-6d %s\n", pid, str(uptr((uint32 *)retval))); works
       * Do we have somekind of auto type resolution available in the future.?
- Now all the deferences without kptr() or uptr() are bpf_probe_read(), [ which just fails on s390x, but this is for good purpose ].

**Comment by fbs** [comment]
>     * Tested few cases on s390x. kptr() and uptr() resolves to bpf_probe_read_kernel() and bpf_probe_read_user, great!!
> 
>     * -v does not seems to show the bpf bytecode. It is shown only during error. This is not relevant to this PR. However, It makes me to use bpftool each time to see the bpf bytecode.
> 

Weird, cannot reproduce this

```
sudo bpftrace -e 'k:f { @=kptr((int8*)123) }' -v
Attaching 1 probe...

Program ID: 2481

Bytecode:
0: (b7) r1 = 0
1: (7b) *(u64 *)(r10 -16) = r1
...
Running...
^C
````

>     * Tried to change bashreadline.bt
>       
>       * printf("%-6d %s\n", pid, str(uptr(retval)));
>       * ERROR: uptr() only supports pointer arguments (integer provided)
>       * printf("%-6d %s\n", pid, str(uptr((uint32 *)retval))); works
>         
>         * Do we have somekind of auto type resolution available in the future.?

I guess we can add it for the `str()` case, internally cast it to a `(char*)`.

**Comment by sumanthkorikkar** [comment]
> > ```
> > * Tested few cases on s390x. kptr() and uptr() resolves to bpf_probe_read_kernel() and bpf_probe_read_user, great!!
> > 
> > * -v does not seems to show the bpf bytecode. It is shown only during error. This is not relevant to this PR. However, It makes me to use bpftool each time to see the bpf bytecode.
> > ```
> 
> Weird, cannot reproduce this
> 
ok. Then I can check this case...
> ```
> sudo bpftrace -e 'k:f { @=kptr((int8*)123) }' -v
> Attaching 1 probe...
> 
> Program ID: 2481
> 
> Bytecode:
> 0: (b7) r1 = 0
> 1: (7b) *(u64 *)(r10 -16) = r1
> ...
> Running...
> ^C
> ```
> 
> > ```
> > * Tried to change bashreadline.bt
> >   
> >   * printf("%-6d %s\n", pid, str(uptr(retval)));
> >   * ERROR: uptr() only supports pointer arguments (integer provided)
> >   * printf("%-6d %s\n", pid, str(uptr((uint32 *)retval))); works
> >     
> >     * Do we have somekind of auto type resolution available in the future.?
> > ```
> 
> I guess we can add it for the `str()` case, internally cast it to a `(char*)`.

ok.

**Comment by fbs** [comment]
I'll merge this on friday (~40h from now) if there are no objections, that way @sumanthkorikkar can start building on this

**Pull request merged**